web/src/message_list_tooltips.ts

```diff
@@ -349,6 +352,64 @@ export function initialize(): void {
         },
     });
 
+    message_list_tooltip(".message_content > p > a[data-message-link-type='external_inline_url']", {
+        onShow(instance) {
+            const $elem = $(instance.reference);
+            const url = $elem.attr("href");
+            instance.setContent(
+                parse_html(render_message_inline_url_tooltip({tooltip_label: url})),
+            );
+        },
+        onHidden(instance) {
+            instance.destroy();
+        },
+        placement: "bottom",
+    });
+
+    message_list_tooltip('.message_content > p > a[data-message-link-type="internal_named_link"]', {
+        delay: LONG_HOVER_DELAY,
+        onShow(instance) {
+            const $elem = $(instance.reference);
+            const url = $elem.attr("href");
+            let tooltip_label;
+            if (url?.startsWith("/user_uploads/")) {
+                // We add the word "download" to make clear what will
+                // happen when clicking the file.  This is particularly
+                // important in the desktop app, where hovering a URL does
+                // not display the URL like it does in the web app.
+                tooltip_label = $t(
+                    {defaultMessage: "Download {filename}"},
+                    {
+                        filename: decodeURIComponent(url.slice(url.lastIndexOf("/") + 1)),
+                    },
+                );
+            } else {
+                // We show the tooltip in canonical form for the
+                // destination internal link is pointing to.
+                let format;
+                const stream_topic_name = hash_util.decode_stream_topic_from_url(
+                    window.location.origin + url!,
+                );
+
+                if (stream_topic_name !== null) {
+                    const channel = stream_data.get_stream_name_from_id(
+                        stream_topic_name.stream_id,
+                    );
+                    const topic = stream_topic_name.topic_name;
+                    format = topic ? `#${channel} > ${topic}` : `#${channel}`;
+                } else {
+                    format = "this";
```

Comment:
"Go to this" doesn't really convey anything and is rather confusing. If we are unable to determine if the internal link is either a user uploaded file, stream or stream+topic at this stage, we should probably just show the full url instead.